packages/core/source/legacy/host.ts

```diff
@@ -0,0 +1,335 @@
+    const parentNode = tree[parentId];
+    if (!parentNode?.[index]) return;
+
+    const id = parentNode[index].id;
+    parentNode.splice(index, 1);
+    cleanupNode(id);
+  }
+
+  function cleanupNode(id: string) {
+    const nodeChildren = tree[id];
+
+    if (nodeChildren) {
+      for (const child of nodeChildren) {
+        cleanupNode(child.id);
+      }
+
+      delete tree[id];
+    }
+  }
+
+  function mutate(records: RemoteMutationRecord[]) {
+    for (const record of records) {
+      const [mutationType, parentId] = record;
+
+      switch (mutationType) {
+        case MUTATION_TYPE_INSERT_CHILD: {
+          const node = record[2];
+          const index = record[3];
+          persistNode(parentId, node, index);
+          break;
+        }
+        case MUTATION_TYPE_REMOVE_CHILD: {
+          const index = record[2];
+          removeNode(parentId, index);
+          break;
+        }
+      }
+    }
+
+    connection.mutate(records);
+  }
+
+  return function remoteChannel<T extends keyof LegacyActionArgumentMap>(
+    type: T,
+    ...payload: LegacyActionArgumentMap[T]
+  ) {
+    switch (type) {
+      case LEGACY_ACTION_MOUNT:
+        const [nodes] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_MOUNT];
+
+        const records = nodes.map(
+          (node, index) =>
+            [
+              MUTATION_TYPE_INSERT_CHILD,
+              ROOT_ID,
+              adaptLegacyNodeSerialization(node, options),
+              index,
+            ] satisfies RemoteMutationRecord,
+        );
+
+        mutate(records);
+
+        break;
+
+      case LEGACY_ACTION_INSERT_CHILD:
+        const [parentId, index, child] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_INSERT_CHILD];
+
+        mutate([
+          [
+            MUTATION_TYPE_INSERT_CHILD,
+            parentId ?? ROOT_ID,
+            adaptLegacyNodeSerialization(child, options),
+            index,
+          ],
+        ]);
+
+        break;
+
+      case LEGACY_ACTION_REMOVE_CHILD: {
+        const [parentID, removeIndex] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_REMOVE_CHILD];
+
+        mutate([
+          [MUTATION_TYPE_REMOVE_CHILD, parentID ?? ROOT_ID, removeIndex],
+        ]);
+
+        break;
+      }
+
+      case LEGACY_ACTION_UPDATE_TEXT: {
+        const [textId, text] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_UPDATE_TEXT];
+
+        mutate([[MUTATION_TYPE_UPDATE_TEXT, textId, text]]);
+
+        break;
+      }
+
+      case LEGACY_ACTION_UPDATE_PROPS: {
+        const [id, props] =
+          payload as LegacyActionArgumentMap[typeof LEGACY_ACTION_UPDATE_PROPS];
+
+        const records = [];
+
+        for (const [key, value] of Object.entries(props)) {
+          if (isFragment(value)) {
+            const index = tree[id]?.findIndex(({slot}) => slot === key) ?? -1;
+
+            if (index !== -1) {
```

Comment:
I'd probably flip this to `index >= 0`, so readers don’t need to handle double negation once they get down to the `else`

Reply:
Oops, I see there is no `else` block for this case, but I still think it's generally nicer to check for truthy conditions in cases like this.

Reply:
`findIndex` is one of the rare cases when I personally prefer to always compare to `-1` as `-1` here holds a special meaning of "not found". I don't mind changing that to `>= 0` though.